openwisp_firmware_upgrader/signals.py

```diff
@@ -0,0 +1,85 @@
+import logging
+
+from django.db import transaction
+
+from .tasks import delete_firmware_files
+
+logger = logging.getLogger(__name__)
+
+
+def _collect_firmware_files(firmware_images):
+    """
+    Collects file paths from firmware images that need to be deleted
+    """
+    paths = []
+    for image in firmware_images:
+        if image.file:
+            paths.append(image.file.name)
+    return paths
+
+
+def delete_build_files(sender, instance, **kwargs):
+    """
+    Delete firmware files when a Build is deleted
+    """
+    logger.info(f"Build delete signal received for build {instance.pk}")
+    firmware_images = instance.firmwareimage_set.all()
```

Comment:
`all()` evaluates the queryset immediately. It is generally preferred to defer query evaluation to as late in code as possible. 

We can further optimize this queryset by only retrieving the file field from the database. 

```
instance.firmwareimage_set.values_list('file', flat=True)
```

This directly returns a list of file paths.